### PR TITLE
Add support for CodeLLDB extension

### PR DESCRIPTION
**Pull request by wolfpld**
Currently, the vscode-meson extension is hardcoded to support only Microsoft's C/C++ extension, which provides the `cppdbg` debugger. However, using the Microsoft extension may not be preferable. An alternative extension, CodeLLDB (https://github.com/vadimcn/codelldb), also provides C/C++ support, including the `lldb` debugger.

These changes add a second debug configuration provider that works with the `lldb` debugger.

These changes also introduce an extension availability check when trying to run tests with debugging. Both the `cppdbg` and `lldb` debuggers are supported, in that order. In addition, a proper error message is now displayed instead of the cryptic "Configured debug type 'cppdbg' is not supported" when no debuggers are available.

Note: The `lldb` debug configuration sets the working directory to `path.dirname(this.path)`, which translates to the parent directory of the build root. This matches the `cppdbg` debug configuration. However, the tests debug configuration falls back to `workspaceRelative(extensionConfiguration("buildFolder")`. Shouldn't this be the same?

### CODE REVIEW

File: src/configprovider-lldb.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    let ret: vscode.DebugConfiguration[] = [];
+
+    for (const target of executables) {
+      if (!target.target_sources?.some((source) => ["cpp", "c"].includes(source.language))) {
+        continue;
+      }
+
+      let debugConfiguration = await this.createDebugConfiguration(target);
+      ret.push({ ...configDebugOptions, ...debugConfiguration });
+    }
```
Comment (tristan957) [inline src/configprovider-lldb.ts:45]: Looks like some of these lets can be const to me.

File: src/tests.ts [hunk 2]
```diff
@@ -99,6 +99,24 @@ export async function testDebugHandler(
...
+    debugType = "cppdbg";
+  } else {
+    const codelldb = vscode.extensions.getExtension("vadimcn.vscode-lldb");
+    if (codelldb && codelldb.isActive) {
+      debugType = "lldb";
+    }
+  }
+
+  if (!debugType) {
+    vscode.window.showErrorMessage("No debugger extension found. Please install one and try again");
```
Comment (tristan957) [inline src/tests.ts:115]: Can you add a period at the end of this message?

**Comment by tristan957** [comment]
Can you rephrase the first commit message? Something like `Be more specific about the current debug provider" or something would work for me.

**Comment by tristan957** [comment]
Also add a changelog entry.

File: src/extension.ts [hunk 3]
```diff
@@ -41,7 +42,15 @@ export async function activate(ctx: vscode.ExtensionContext) {
...
       "cppdbg",
-      new DebugConfigurationProvider(buildDir),
+      new DebugConfigurationProviderCppdbg(buildDir),
+      vscode.DebugConfigurationProviderTriggerKind.Dynamic,
+    ),
+  );
+
+  ctx.subscriptions.push(
+    vscode.debug.registerDebugConfigurationProvider(
+      "lldb",
```
Comment (tristan957) [inline src/extension.ts:52]: Can you expose these strings as static public constants on the DebugConfigurationProvider implementations?

File: src/configprovider-lldb.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import * as vscode from "vscode";
+import * as path from "path";
+import { getMesonTargets } from "./introspection";
+import { Target } from "./types";
+import { extensionConfiguration, getTargetName } from "./utils";
+
+export class DebugConfigurationProviderLldb implements vscode.DebugConfigurationProvider {
```
Comment (tristan957) [inline src/configprovider-lldb.ts:7]: How much of this class is now a copy-paste of the Cppdbg one? Anyway that we could factor the copy-paste into a common abstract base class like "MesonDebugConfigurationProvider"?

**Review by tristan957 — COMMENTED** [review]
I'll spin a release as soon as I get the requested changes. Thanks for this!

File: src/tests.ts (hunk 2 above)
Comment (wolfpld) [reply src/tests.ts:115]: There is no period at the end in other error messages in this file. In other places in the code, there's no consistency in how this is handled. I can do this, but fixing it seems to be outside the scope of this PR.

https://github.com/mesonbuild/vscode-meson/blob/2d91eebf3beebfd9429e098d8f67cac1ab88ea7d/src/tests.ts#L54
https://github.com/mesonbuild/vscode-meson/blob/2d91eebf3beebfd9429e098d8f67cac1ab88ea7d/src/tests.ts#L97

Comment (tristan957) [reply src/tests.ts:115]: Yeah keep it as is then. Thanks for checking.

File: src/extension.ts (hunk 3 above)
Comment (wolfpld) [reply src/extension.ts]: Should be fixed by f17751a.

File: src/configprovider-cppdbg.ts [hunk 5]
```diff
@@ -10,6 +10,8 @@ export enum MIModes {
 }
 
 export class DebugConfigurationProviderCppdbg implements vscode.DebugConfigurationProvider {
+  static readonly type = "cppdbg";
```
Comment (tristan957) [inline src/configprovider-cppdbg.ts:13]: Rename it to `name`. Thanks for this commit.

Comment (wolfpld) [reply src/configprovider-cppdbg.ts:13]: Static property 'name' conflicts with built-in property 'Function.name' of constructor function 'DebugConfigurationProviderCppdbg'.

Comment (tristan957) [reply src/configprovider-cppdbg.ts:13]: type is fine then!

File: src/configprovider-lldb.ts (hunk 4 above)
Comment (wolfpld) [reply src/configprovider-lldb.ts]: Quite a lot apparently. e817bb1 should fix things.

File: src/configprovider-lldb.ts (hunk 1 above)
Comment (wolfpld) [reply src/configprovider-lldb.ts]: This was in the code previously. I fixed it in e817bb1.

File: src/configprovider-cppdbg.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import * as vscode from "vscode";
+import { Target } from "./types";
+import { MesonDebugConfigurationProvider } from "./configprovider";
+
+export enum MIModes {
```
Comment (tristan957) [inline src/configprovider-cppdbg.ts:5]: Does this need to be exported anymore?

**Comment by tristan957** [comment]
I think you are missing the MesonDebugConfigurationProvider.

File: src/configprovider-cppdbg.ts (hunk 6 above)
Comment (wolfpld) [reply src/configprovider-cppdbg.ts:5]: I don't know. It was exported before and I don't know the language.

File: src/configprovider.ts [hunk 7]
```diff
@@ -4,23 +4,22 @@ import { getMesonTargets } from "./introspection";
...
+  abstract createDebugConfiguration(target: Target): Promise<vscode.DebugConfiguration>;
+
   async createBaseDebugConfiguration(target: Target): Promise<vscode.DebugConfiguration> {
     const targetName = await getTargetName(target);
+    const name = this.getName();
     return {
-      type: "cppdbg",
-      name: `Debug ${target.name}`,
+      type: name,
+      name: `Debug ${target.name} (` + name + `)`,
```
Comment (tristan957) [inline src/configprovider.ts:22]: Use `${name}`

File: src/configprovider.ts [hunk 8]
```diff
@@ -70,15 +44,7 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
-      let debugConfiguration = null;
-      if (target.target_sources.some((source) => ["cl"].includes(source.compiler[0]))) {
-        debugConfiguration = await this.createMSVCDebugConfiguration(target);
-      } else if (target.target_sources.some((source) => ["cc", "clang"].includes(source.compiler[0]))) {
-        debugConfiguration = await this.createLLDBDebugConfiguration(target);
-      } else {
-        debugConfiguration = await this.createGDBDebugConfiguration(target);
-      }
-
+      let debugConfiguration = await this.createDebugConfiguration(target);
```
Comment (tristan957) [inline src/configprovider.ts:47]: This can also be const right?

File: src/extension.ts [hunk 9]
```diff
@@ -38,13 +39,18 @@ export async function activate(ctx: vscode.ExtensionContext) {
...
 
-  ctx.subscriptions.push(
-    vscode.debug.registerDebugConfigurationProvider(
-      "cppdbg",
-      new DebugConfigurationProvider(buildDir),
-      vscode.DebugConfigurationProviderTriggerKind.Dynamic,
-    ),
-  );
+  const providers = [DebugConfigurationProviderCppdbg, DebugConfigurationProviderLldb];
+
```
Comment (tristan957) [inline src/extension.ts:43]: Remove this newline

File: src/configprovider.ts [hunk 10]
```diff
@@ -4,23 +4,22 @@ import { getMesonTargets } from "./introspection";
...
+  private readonly path: string;
 
   constructor(path: string) {
     this.path = path;
   }
 
+  abstract getName(): string;
+  abstract createDebugConfiguration(target: Target): Promise<vscode.DebugConfiguration>;
+
   async createBaseDebugConfiguration(target: Target): Promise<vscode.DebugConfiguration> {
```
Comment (tristan957) [inline src/configprovider.ts:17]: I would change this function name to `createDebugConfiguration`. Make it not abstract, and use super in the Cppdbg class to access it after overriding it. Does that make sense?

File: src/configprovider-cppdbg.ts (hunk 6 above)
Comment (wolfpld) [reply src/configprovider-cppdbg.ts:5]: The documentation for this is at https://code.visualstudio.com/docs/cpp/launch-json-reference#_mimode and it's just a string that can have two values set.

TBH, I see zero value in having this export here. It just makes you do more reading and writing, so it's actually a net negative. I will remove it and just set the raw strings where appropriate.

```ts
export enum MIModes {
  lldb = "lldb",
  gdb = "gdb",
}
```

**Comment by wolfpld** [comment]
I have addressed all of your suggestions.

File: src/configprovider-cppdbg.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+  async createLLDBDebugConfiguration(target: Target): Promise<vscode.DebugConfiguration> {
+    let debugConfig = await super.createDebugConfiguration(target);
+    debugConfig["MIMode"] = "lldb";
+    return debugConfig;
+  }
+
+  async createMSVCDebugConfiguration(target: Target): Promise<vscode.DebugConfiguration> {
+    let debugConfig = await super.createDebugConfiguration(target);
+    debugConfig.type = "cppvsdbg";
```
Comment (wolfpld) [inline src/configprovider-cppdbg.ts:47]: This line goes against the whole design, but it was here before. I wonder if this even works, considering that you have to specify each of the debugger types to be handled by the configuration provider, in `package.json`.

Comment (tristan957) [reply src/configprovider-cppdbg.ts:47]: Would you mind submitting another PR to refactor this after we get this PR in?

Comment (wolfpld) [reply src/configprovider-cppdbg.ts:47]: I'm not interested in supporting Microsoft's extension, sorry.

**Comment by tristan957** [comment]
I will spin up a release today. Appreciate your work on this.
